cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -109,13 +119,124 @@ func (spm *KubeStaticPodPathManager) BackupManifestDir() string {
+
+	// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+	// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+	// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+	// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+	if err := waiter.WaitForStaticPodControlPlaneHashChange(cfg.NodeName, component, beforePodHash); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Wait for the static pod component to come up and register itself as a mirror pod
+	if err := waiter.WaitForPodsWithLabel("component=" + component); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	return nil
+}
+
+func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, recoverManifests map[string]string) error {
+	// Add etcd static pod spec only if external etcd is not configured
+	if len(cfg.Etcd.Endpoints) != 0 {
+		return fmt.Errorf("external etcd cannot be upgraded with kubeadm")
+	}
+	// Checking health state of etcd before proceeding with the upgrtade
+	etcdStatus, err := util.GetEtcdClusterStatus()
+	if err != nil {
+		return fmt.Errorf("etcd cluster is not healthy: %v", err)
+	}
+
+	// Backing up etcd data store
+	backupEtcdDir := pathMgr.BackupEtcdDir()
+	runningEtcdDir := cfg.Etcd.DataDir
+	if err := util.CopyDir(runningEtcdDir, backupEtcdDir); err != nil {
+		return fmt.Errorf("fail to back up etcd data with %v", err)
+	}
+
+	// Need to check currently used version and version from constants, if differs then uprgade
+	currentEtcdVersion := etcdStatus.Version
+	desiredEtcdVersion, err := constants.EtcdSupportedVersion(cfg.KubernetesVersion)
+	if err != nil {
+		return rollbackEtcdData(cfg, fmt.Errorf("failed to parse the desired etcd version(%s): %v", desiredEtcdVersion.String(), err), pathMgr)
+	}
+
+	// Comparing current etcd version with desired to catch the same version or downgrade condition and fail on them.
+	res, _ := desiredEtcdVersion.Compare(currentEtcdVersion)
+	switch res {
+	case -1:
+		return fmt.Errorf("the requested etcd version (%s) for Kubernetes v(%s) is lower than the currently running version (%s)", desiredEtcdVersion, cfg.KubernetesVersion, currentEtcdVersion)
+	case 0:
+		return fmt.Errorf("the requested etcd version (%s) for Kubernetes v(%s) is the same as the currently running version (%s)", desiredEtcdVersion, cfg.KubernetesVersion, currentEtcdVersion)
+	}
+
+	beforeEtcdPodHash, err := waiter.WaitForStaticPodSingleHash(cfg.NodeName, constants.Etcd)
+	if err != nil {
+		return rollbackEtcdData(cfg, err, pathMgr)
```

Comment:
no rollback should be performed here. fatal case

Reply:
done